contrib/jms/src/main/java/com/google/cloud/dataflow/contrib/jms/JmsIO.java

```diff
@@ -0,0 +1,532 @@
+            private final Coder<T> coder;
+
+            private Bound(Coder<T> coder) {
+                this(null, null, null, null, coder);
+            }
+
+            private Bound(String name, ConnectionFactory connectionFactory,
+                          String queue, String topic, Coder<T> coder) {
+                super(name);
+                this.connectionFactory = connectionFactory;
+                this.queue = queue;
+                this.topic = topic;
+                this.coder = coder;
+            }
+
+            public Bound<T> named(String name) {
+                return new Bound<T>(name,
+                        connectionFactory, queue, topic, coder);
+            }
+
+            public Bound<T> connectionFactory(ConnectionFactory connectionFactory) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder);
+            }
+
+            public Bound<T> queue(String queue) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder);
+            }
+
+            public Bound<T> topic(String topic) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder);
+            }
+
+            public <X> Bound<X> withCoder(Coder<X> coder) {
+                return new Bound<>(name, connectionFactory, queue, topic, coder);
+            }
+
+            @Override
+            public PDone apply(PCollection<T> input) {
+                if (connectionFactory == null) {
+                    throw new java.lang.IllegalStateException("need to set the JMS " +
+                            "connection factory of a " +
+                            "JmsIO.Write transform");
+                }
+                if (queue == null && topic == null) {
+                    throw new java.lang.IllegalStateException("need to set the JMS " +
+                            "destination (queue or topic) of " +
+                            "a JmsIO.Write transform");
+                }
+                input.apply(ParDo.of(new JmsWriter()));
+                return PDone.in(input.getPipeline());
+            }
+
+            @Override
+            protected Coder<T> getDefaultOutputCoder() {
+                return coder;
+            }
+
+            public ConnectionFactory getConnectionFactory() {
+                return connectionFactory;
+            }
+
+            public String getQueue() {
+                return queue;
+            }
+
+            public String getTopic() {
+                return topic;
+            }
+
+            public Coder<T> getCoder() {
+                return coder;
+            }
+
+            private class JmsWriter extends DoFn<T, Void> {
+                private transient Connection connection;
+                private transient Session session;
+
+                @Override
+                public void startBundle(Context c) throws Exception {
+                    connection = connectionFactory.createConnection();
```

Comment:
A few motivating questions:

* do you expect to use this Write for batch jobs or streaming jobs or both?
* how expensive is it to create a connection?
* what resources does each connection use, and are they likely to run into machine-local resource limits?

In batch pipelines, there may be few, very large bundles. So it makes sense to open a connection at the beginning and close it at the end.

However, in streaming pipelines, there are many small, short-lived bundles (very few elements, even 1, per bundle!). If opening a connection takes a long time, this can add a lot of overhead to a pipeline. Additionally, there can be several hundred threads each with an active bundle (so, several hundred connections open) so it's important to think about resource limits.